Add support for conditional access based on roles. Part one.

splatch opened the pull request:
This commit addresses issue #579. It does not solve it fully, but creates a base for security mechanism which covers main entrypoints to Eclipse SmartHome - meaning REST and other servlet based interactions such icons and charts.

While this is just beginning there is lots of things to be reviewed and validated. Initial review of code and its execution have been done by @svilenvul with positive feedback. Issues found by him were addressed in fork.

splatch commented:
FYI - I've signed CLA long time ago using lukasz@dywicki.pl mail, however ip-validation bot does not pick up signed off tag value.

kaikreuzer commented:
No worries. I have manually checked your ECA and can confirm that it is valid. This PR will anyhow have to go through IP team approval, so just ignore the IP validation failure here.

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/internal/auth/DefaultAuthenticationManager.java:
```diff
@@ -0,0 +1,75 @@
+ *
+ */
+@Component
+public class DefaultAuthenticationManager implements AuthenticationManager {
```
If there is only a single implementation which we provide, we usually rather call it `AuthenticationManagerImpl`.

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/internal/auth/DefaultAuthenticationManager.java:
```diff
@@ -0,0 +1,75 @@
+        }
+
+        if (unmatched) {
+            throw new UnsupportedCredentialsException("Unsupported credentials specified " + credentials);
```
You should not include the credentials here (as this exception might get logged), only their class.

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/internal/auth/DefaultAuthenticationManager.java:
```diff
@@ -0,0 +1,75 @@
+        if (unmatched) {
+            throw new UnsupportedCredentialsException("Unsupported credentials specified " + credentials);
+        }
+        throw new AuthenticationException("Could not authenticate credentials " + credentials);
```
You should not include the credentials here (as this exception might get logged), only their class.

kaikreuzer commented on bundles/io/org.eclipse.smarthome.io.http.core/.project:
```diff
@@ -0,0 +1,33 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<projectDescription>
+	<name>org.eclipse.smarthome.io.http.core</name>
```
We already have a `org.eclipse.smarthome.io.http` bundle, so this somehow duplicates it.
I think we could easily move the classes of this bundle to a package `org.eclipse.smarthome.io.http.servlet` in the `org.eclipse.smarthome.io.http` bundle and remove this one here.

splatch replied:
I messed up dot files (I have global git ignore to not commit them by default) and then I been adding them manually on other machine. Will fix.

kaikreuzer commented on bundles/io/org.eclipse.smarthome.io.http.auth.form/src/main/java/org/eclipse/smarthome/io/http/auth/form/internal/LoginService.java:
```diff
@@ -0,0 +1,74 @@
+@Component
+public class LoginService {
+
+    private static final String LOGIN_FORM_ADDRESS = "/login/form";
```
This should be "/login", so that the page is available as `/login/index.html`.
It might even be more elegant to register a servlet under `/login`, which simply serves the page.

kaikreuzer replied:
Just noticed that the `?process` is anyhow not processed by the form, so I assume you'll have to make this a servlet anyhow, so that you can set the Basic Auth HTTP header, right?

kaikreuzer commented on bundles/io/org.eclipse.smarthome.io.http.auth.form/web/index.html:
```diff
@@ -0,0 +1,28 @@
+<body>
+
+<div id="login_form">
+    <form name="login-form" method="post" action="/login/form/index.html?process" id="f1">
```
Better don't use an absolute, but a relative url here - `index.html?process` should just be fine.

kaikreuzer commented on bundles/io/org.eclipse.smarthome.io.http.auth/src/main/java/org/eclipse/smarthome/io/http/auth/internal/AuthenticationHandler.java:
```diff
@@ -0,0 +1,160 @@
+
+    // configuration properties
+    private boolean enabled = true;
+    private String loginUri = "/login";
```
Use a url that works in ESH out of the box without requiring any configuration - e.g. "/login/index.html", if this is where we'd register the form.

kaikreuzer commented on bundles/io/org.eclipse.smarthome.io.http.auth/src/main/java/org/eclipse/smarthome/io/http/auth/internal/AuthenticationHandler.java:
```diff
@@ -0,0 +1,160 @@
+
+        if (error instanceof AuthenticationException) {
+            // force client redirect
+            response.setHeader("Location", loginUri);
```
Shouldn't you pass the called url as a parameter, so that the form can actually redirect back once the user authenticated successfully?

splatch replied:
Correct, will update code to support such scenario.

splatch commented on bundles/auth/org.eclipse.smarthome.auth.password/src/main/java/org/eclipse/smarthome/auth/password/UsernamePasswordCredentials.java:
```diff
@@ -10,7 +10,9 @@
  * SPDX-License-Identifier: EPL-2.0
  */
-package org.eclipse.smarthome.core.auth;
+package org.eclipse.smarthome.auth.password;
```
I moved this away as I can think of use cases where password authentication is not one supported by default. Also moving credential to separate packages causes that extractors working with password based access are free of dependency to smarthome core.

kaikreuzer replied:
Defining that there is such a thing as password credentials does not mean that there is support for it - the support only comes by including the http.auth.basic bundle.

> are free of dependency to smarthome core.

Well, smarthome.core.auth contains the basic building blocks, and it is expected to be the root dependency for every auth feature - I don't think this is therefore an argument.

splatch commented on bundles/io/org.eclipse.smarthome.io.http.core/src/main/java/org/eclipse/smarthome/io/http/core/BaseSmartHomeServlet.java:
```diff
@@ -0,0 +1,68 @@
+
+    protected void activate(String alias, HttpContext httpContext) {
+        try {
+            logger.debug("Starting up chart servlet at " + alias);
```
You got me. This is copy paste.

splatch commented:
@kaikreuzer Thank you very much for review and kind words. 

> I couldn't get the code to work (in a sense that I'd been able to login), but afaics, this is simply not yet fully implemented, correct?

I found that sometimes I had to refresh or restart some bundles in order to get all services working together. Code which is currently implemented offers only very thin protection layer, namely a guard, and does not populate nor keep session. This means that even if you pass authentication with basic credentials - you still might get redirected to login page on next click. This is because there is no session, and no cookie returned to client so auth result is simply lost over time. In order to keep login operation result somewhat more persistent we need to co come with local JWT representing `Authentication`. If we will have that we could use a cookie to store a token until all UIs support authentication of out the box.
I didn't reach this place yet because it touches a bit of authorization planned for second phase. To start this we need to define JWT claim names (or just single name we want to use for keeping role) and their values.

> So what's your plan for the next steps? I think I can (at least from my perspective) give you green light for what you have done so far and the direction you are heading - I'd therefore love to see a working scenario with Basic Auth very soon and one with JWT shortly after :-)

I used so far form and basic login to verify if entire idea works. My main intention was to provide html form for human interactions (to not rely on basic auth) and leave HTTP basic for calls from various shell scripts. Both basic and form credential extractors could have additional configuration to restrict what client addresses they do accept. By this a access based on HTTP basic, which is extremely easy to tap, could be restricted just to localhost. 
Basic auth should work already. I didn't plan yet to work on JWT due to reasons listed above. 

I tested entire thing with openhab distro and it have worked. I will update code, retest and update PR.

kaikreuzer commented:
> I tested entire thing with openhab distro and it have worked.

Well, we probably have a different definition of "working" then :-)
Yes, all bundles resolved and services started and all hooks worked indeed - so what is there is working as expected, I agree.
It is just that the parts do not yet interact in a way that it brings any working feature to the user: The redirect to the login form fails due to a wrong default url, the submission of form data is not at all processed and thus the user is never authenticated, i.e. cannot access any page. And as you say, even if that would work, due to a lacking session, it wouldn't be usable.

I agree that BasicAuth is not the way to go, but as long as we are working on JWT support, it might be an idea to change the login form to ask the browser to do BasicAuth (which it will keep sending in further requests, if I am not mistaken). This way, we could at least have a working prototype. Wdyt?

Wrt JWT claim names: I have no experience in how to choose them and what makes sense there. Do you have some references and could come up with a suggestion? @JochenHiller Do you have any wishes/suggestions here?

wborn commented on bundles/io/org.eclipse.smarthome.io.http.auth.basic/pom.xml:
```diff
@@ -0,0 +1,22 @@
+  <properties>
+    <bundle.symbolicName>org.eclipse.smarthome.io.http.auth.basic</bundle.symbolicName>
+    <bundle.namespace>org.eclipse.smarthome.io.http.auth.basic</bundle.namespace>
+  </properties>
```
Please remove these properties. See #6060.

wborn commented on bundles/io/org.eclipse.smarthome.io.http.auth.form/pom.xml:
```diff
@@ -0,0 +1,22 @@
+  <properties>
+    <bundle.symbolicName>org.eclipse.smarthome.io.http.auth.form</bundle.symbolicName>
+    <bundle.namespace>org.eclipse.smarthome.io.http.auth.form</bundle.namespace>
+  </properties>
```
Please remove these properties. See #6060.

wborn commented on bundles/io/org.eclipse.smarthome.io.http.auth.jwt/pom.xml:
```diff
@@ -0,0 +1,22 @@
+  <properties>
+    <bundle.symbolicName>org.eclipse.smarthome.io.http.auth.jwt</bundle.symbolicName>
+    <bundle.namespace>org.eclipse.smarthome.io.http.auth.jwt</bundle.namespace>
+  </properties>
```
Please remove these properties. See #6060.

wborn commented on bundles/io/org.eclipse.smarthome.io.http.auth/pom.xml:
```diff
@@ -0,0 +1,22 @@
+  <properties>
+    <bundle.symbolicName>org.eclipse.smarthome.io.http.auth</bundle.symbolicName>
+    <bundle.namespace>org.eclipse.smarthome.io.http.auth</bundle.namespace>
+  </properties>
```
Please remove these properties. See #6060.

wborn commented on bundles/io/org.eclipse.smarthome.io.http.core/pom.xml:
```diff
@@ -0,0 +1,22 @@
+  <properties>
+    <bundle.symbolicName>org.eclipse.smarthome.io.http.core</bundle.symbolicName>
+    <bundle.namespace>org.eclipse.smarthome.io.http.core</bundle.namespace>
+  </properties>
```
Please remove these properties. See #6060.

wborn commented on bundles/io/org.eclipse.smarthome.io.http.test/pom.xml:
```diff
@@ -0,0 +1,39 @@
+  <properties>
+    <bundle.symbolicName>org.eclipse.smarthome.io.http.test</bundle.symbolicName>
+    <bundle.namespace>org.eclipse.smarthome.io.http.test</bundle.namespace>
+  </properties>
```
Please remove these properties. See #6060.

wborn commented on bundles/io/org.eclipse.smarthome.io.http/pom.xml:
```diff
@@ -0,0 +1,22 @@
+  <properties>
+    <bundle.symbolicName>org.eclipse.smarthome.io.http</bundle.symbolicName>
+    <bundle.namespace>org.eclipse.smarthome.io.http</bundle.namespace>
+  </properties>
```
Please remove these properties. See #6060.

wborn commented on bundles/io/org.eclipse.smarthome.io.rest.optimize/pom.xml:
```diff
@@ -14,4 +14,10 @@
 
   <name>Eclipse SmartHome REST JAX-RS Optimizations</name>
+
+  <properties>
```
That's right. See #6060.

wborn commented on bundles/ui/org.eclipse.smarthome.ui.login/pom.xml:
```diff
@@ -0,0 +1,23 @@
+  <properties>
+    <bundle.symbolicName>org.eclipse.smarthome.ui.login</bundle.symbolicName>
+    <bundle.namespace>org.eclipse.smarthome.ui.login</bundle.namespace>
+  </properties>
```
Please remove these properties. See #6060.

kaikreuzer commented on bundles/io/org.eclipse.smarthome.io.rest.auth/src/main/java/org/eclipse/smarthome/io/rest/auth/AbstractSecurityHandler.java:
```diff
@@ -27,10 +27,14 @@
 /**
  * Base type for handling authentication and authorization delegated from rest/jaxrs service layer.
  *
+ * @deprecated Since introduction of security on lower layer (servlets), this extension which is
```
I'd be ok with completely removing this class - I doubt that anyone is using this as it never fully worked.

splatch commented on bundles/io/org.eclipse.smarthome.io.rest.optimize/src/main/java/org/eclipse/smarthome/io/rest/optimize/internal/SmartHomeRestServletConfig.java:
```diff
@@ -0,0 +1,57 @@
+/**
+ * Custom servlet configuration for jaxrs handler.
+ *
+ * This extension doesn't do much, just forces usage of http context created somewhere else.
```
Yes, this is and extension point to jaxrs publisher which allows to customize OSGi servlet registration and bind it to specific http context instead of default one. Thanks to this simple hook we are able to link all jaxrs endpoints to logical ESH-specific security domain.

splatch commented:
>> tested entire thing with openhab distro and it have worked.

> Well, we probably have a different definition of "working" then :-)

First goal which was a must was a system wide protection switch which will enable security on all servlets and that was my primary goal in submitting this pull request. I agree that it still has little or no value for end user, yet going further requires touching bits of JWT which was decided to be used as session mechanism.

> Yes, all bundles resolved and services started and all hooks worked indeed - so what is there is working as expected, I agree.
> It is just that the parts do not yet interact in a way that it brings any working feature to the user: The redirect to the login form fails due to a wrong default url, the submission of form data is not at all processed and thus the user is never authenticated, i.e. cannot access any page. And as you say, even if that would work, due to a lacking session, it wouldn't be usable.

I updated code to satisfy this requirement - once user enters under protected resource it is redirected to login method selection where he can just select a form. After successful authentication he is redirected back to place where he started whole process, see RedirectHandler which has been added in one of last commits. Still, even if user is properly authenticated he is forwarded back to login because authentication result is not persisted.

> I agree that BasicAuth is not the way to go, but as long as we are working on JWT support, it might be an idea to change the login form to ask the browser to do BasicAuth (which it will keep sending in further requests, if I am not mistaken). This way, we could at least have a working prototype. Wdyt?

Yes, this gonna solve a cyclic redirect. I will add a basic prompt for handling that.

> Wrt JWT claim names: I have no experience in how to choose them and what makes sense there. Do you have some references and could come up with a suggestion? @JochenHiller Do you have any wishes/suggestions here?

Based on my own knowledge of JWT spec I can't find any. We know it need to be short, yet unique so it doesn't conflict with any [standard claim](https://www.iana.org/assignments/jwt/jwt.xhtml). We can use `esh_role` containing a JSON list for now.
Main reason why I am asking about this is because there is more work on that part to be made. We need to generate (or reuse) key pair, create token with required claims and also validate it once client sends it back. Because implementing even fragment of JWT specification requires significant amount of work I usually used 3rd party library for that. I am not sure what is your take on that since it will be an external dependency.

kaikreuzer commented on bundles/io/org.eclipse.smarthome.io.rest.auth/META-INF/MANIFEST.MF:
```diff
@@ -10,9 +10,11 @@ Bundle-Version: 0.10.0.qualifier
  org.eclipse.smarthome.io.rest.auth,
+ org.osgi.service.http,
  org.slf4j
 Automatic-Module-Name: org.eclipse.smarthome.io.rest.auth
```
```Service-Component: OSGI-INF/*.xml```
is missing here

kaikreuzer commented on bundles/io/org.eclipse.smarthome.io.rest.auth/META-INF/MANIFEST.MF:
```diff
@@ -10,9 +10,11 @@ Bundle-Version: 0.10.0.qualifier
  javax.ws.rs.container,
  org.eclipse.jdt.annotation;resolution:=optional,
  org.eclipse.smarthome.core.auth,
  org.eclipse.smarthome.io.rest.auth,
```
remove

kaikreuzer commented on bundles/ui/org.eclipse.smarthome.ui.login/src/main/java/org/eclipse/smarthome/ui/login/internal/LoginMethodsServlet.java:
```diff
@@ -0,0 +1,91 @@
+public class LoginMethodsServlet extends SmartHomeServlet {
+
+    private static final String REDIRECT_PARAM_NAME = "redirect";
+    private static final String LOGIN_ADDRESS = "/login";
```
I get a 403 when accessing this URI...

kaikreuzer commented on bundles/io/org.eclipse.smarthome.io.http.auth.form/src/main/java/org/eclipse/smarthome/io/http/auth/form/internal/FormService.java:
```diff
@@ -0,0 +1,75 @@
+@Component
+public class FormService {
+
+    static final String LOGIN_FORM_ADDRESS = "/login/form";
```
You already register the servlet under `/login` - doesn't this cause conflicts here?

kaikreuzer commented on extensions/ui/org.eclipse.smarthome.ui.classic/OSGI-INF/webappservlet.xml:
```diff
@@ -0,0 +1,29 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
please set these xmls to ignore

kaikreuzer commented on bundles/ui/org.eclipse.smarthome.ui.login/.classpath:
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Again, please let us try to keep the number of additional bundles small: This one does not have any too specific dependencies and consists out of a single class only - this should imho be moved to the smarthome.ui bundle.

kaikreuzer commented:
> Because implementing even fragment of JWT specification requires significant amount of work I usually used 3rd party library for that. I am not sure what is your take on that since it will be an external dependency.

I agree that we should not reimplement such stuff ourselves. I'd hope that there are some lean libs available, do you have any specific one in mind already?

splatch commented:
> I agree that we should not reimplement such stuff ourselves. I'd hope that there are some lean libs available, do you have any specific one in mind already?

I've used nimbusds so far and it worked quite fine with most of use cases. @antoniosv in antoniosv/smarthome#1 used  jose4j for that which might be also a good alternative.

kaikreuzer commented:
Both look fine to me wrt size and license.

wborn commented on bundles/auth/org.eclipse.smarthome.auth.jwt/.settings/org.eclipse.jdt.core.prefs:
```diff
@@ -0,0 +1,8 @@
+org.eclipse.jdt.core.compiler.problem.assertIdentifier=error
+org.eclipse.jdt.core.compiler.problem.enumIdentifier=error
+org.eclipse.jdt.core.compiler.problem.forbiddenReference=warning
+org.eclipse.jdt.core.compiler.source=1.8
```
Please remove this file because it causes the default Eclipse project settings to be overridden. See also #6106.

wborn commented on bundles/auth/org.eclipse.smarthome.auth.jwt/.settings/org.maven.ide.eclipse.prefs:
```diff
@@ -0,0 +1,9 @@
+resolveWorkspaceProjects=true
+resourceFilterGoals=process-resources resources\:testResources
+skipCompilerPlugin=true
+version=1
```
Please remove this file because it causes the default Eclipse project settings to be overridden. See also #6106.

wborn commented on bundles/io/org.eclipse.smarthome.io.http.auth.basic/.settings/org.eclipse.jdt.core.prefs:
```diff
@@ -0,0 +1,8 @@
+org.eclipse.jdt.core.compiler.problem.assertIdentifier=error
+org.eclipse.jdt.core.compiler.problem.enumIdentifier=error
+org.eclipse.jdt.core.compiler.problem.forbiddenReference=warning
+org.eclipse.jdt.core.compiler.source=1.8
```
Please remove this file because it causes the default Eclipse project settings to be overridden. See also #6106.

wborn commented on bundles/io/org.eclipse.smarthome.io.http.auth.basic/.settings/org.maven.ide.eclipse.prefs:
```diff
@@ -0,0 +1,9 @@
+resolveWorkspaceProjects=true
+resourceFilterGoals=process-resources resources\:testResources
+skipCompilerPlugin=true
+version=1
```
Please remove this file because it causes the default Eclipse project settings to be overridden. See also #6106.

wborn commented on bundles/io/org.eclipse.smarthome.io.http.auth.form/.settings/org.eclipse.jdt.core.prefs:
```diff
@@ -0,0 +1,8 @@
+org.eclipse.jdt.core.compiler.problem.assertIdentifier=error
+org.eclipse.jdt.core.compiler.problem.enumIdentifier=error
+org.eclipse.jdt.core.compiler.problem.forbiddenReference=warning
+org.eclipse.jdt.core.compiler.source=1.8
```
Please remove this file because it causes the default Eclipse project settings to be overridden. See also #6106.

wborn commented on bundles/io/org.eclipse.smarthome.io.http.auth.form/.settings/org.maven.ide.eclipse.prefs:
```diff
@@ -0,0 +1,9 @@
+resolveWorkspaceProjects=true
+resourceFilterGoals=process-resources resources\:testResources
+skipCompilerPlugin=true
+version=1
```
Please remove this file because it causes the default Eclipse project settings to be overridden. See also #6106.

wborn commented on bundles/io/org.eclipse.smarthome.io.http.auth.jwt/.settings/org.eclipse.jdt.core.prefs:
```diff
@@ -0,0 +1,8 @@
+org.eclipse.jdt.core.compiler.problem.assertIdentifier=error
+org.eclipse.jdt.core.compiler.problem.enumIdentifier=error
+org.eclipse.jdt.core.compiler.problem.forbiddenReference=warning
+org.eclipse.jdt.core.compiler.source=1.8
```
Please remove this file because it causes the default Eclipse project settings to be overridden. See also #6106.

wborn commented on bundles/io/org.eclipse.smarthome.io.http.auth.jwt/.settings/org.maven.ide.eclipse.prefs:
```diff
@@ -0,0 +1,9 @@
+resolveWorkspaceProjects=true
+resourceFilterGoals=process-resources resources\:testResources
+skipCompilerPlugin=true
+version=1
```
Please remove this file because it causes the default Eclipse project settings to be overridden. See also #6106.

wborn commented on bundles/io/org.eclipse.smarthome.io.http.auth/.settings/org.eclipse.jdt.core.prefs:
```diff
@@ -0,0 +1,8 @@
+org.eclipse.jdt.core.compiler.problem.assertIdentifier=error
+org.eclipse.jdt.core.compiler.problem.enumIdentifier=error
+org.eclipse.jdt.core.compiler.problem.forbiddenReference=warning
+org.eclipse.jdt.core.compiler.source=1.8
```
Please remove this file because it causes the default Eclipse project settings to be overridden. See also #6106.

wborn commented on bundles/io/org.eclipse.smarthome.io.http.auth/.settings/org.maven.ide.eclipse.prefs:
```diff
@@ -0,0 +1,9 @@
+resolveWorkspaceProjects=true
+resourceFilterGoals=process-resources resources\:testResources
+skipCompilerPlugin=true
+version=1
```
Please remove this file because it causes the default Eclipse project settings to be overridden. See also #6106.

wborn commented on bundles/io/org.eclipse.smarthome.io.http.test/.settings/org.eclipse.jdt.core.prefs:
```diff
@@ -0,0 +1,8 @@
+org.eclipse.jdt.core.compiler.problem.assertIdentifier=error
+org.eclipse.jdt.core.compiler.problem.enumIdentifier=error
+org.eclipse.jdt.core.compiler.problem.forbiddenReference=warning
+org.eclipse.jdt.core.compiler.source=1.8
```
Please remove this file because it causes the default Eclipse project settings to be overridden. See also #6106.

wborn commented on bundles/io/org.eclipse.smarthome.io.http.test/.settings/org.maven.ide.eclipse.prefs:
```diff
@@ -0,0 +1,9 @@
+resolveWorkspaceProjects=true
+resourceFilterGoals=process-resources resources\:testResources
+skipCompilerPlugin=true
+version=1
```
Please remove this file because it causes the default Eclipse project settings to be overridden. See also #6106.

wborn commented on bundles/io/org.eclipse.smarthome.io.http/.settings/org.eclipse.jdt.core.prefs:
```diff
@@ -0,0 +1,8 @@
+org.eclipse.jdt.core.compiler.problem.assertIdentifier=error
+org.eclipse.jdt.core.compiler.problem.enumIdentifier=error
+org.eclipse.jdt.core.compiler.problem.forbiddenReference=warning
+org.eclipse.jdt.core.compiler.source=1.8
```
Please remove this file because it causes the default Eclipse project settings to be overridden. See also #6106.

wborn commented on bundles/io/org.eclipse.smarthome.io.http/.settings/org.maven.ide.eclipse.prefs:
```diff
@@ -0,0 +1,9 @@
+resolveWorkspaceProjects=true
+resourceFilterGoals=process-resources resources\:testResources
+skipCompilerPlugin=true
+version=1
```
Please remove this file because it causes the default Eclipse project settings to be overridden. See also #6106.

wborn commented on bundles/ui/org.eclipse.smarthome.ui.login/.settings/org.eclipse.jdt.core.prefs:
```diff
@@ -0,0 +1,8 @@
+org.eclipse.jdt.core.compiler.problem.assertIdentifier=error
+org.eclipse.jdt.core.compiler.problem.enumIdentifier=error
+org.eclipse.jdt.core.compiler.problem.forbiddenReference=warning
+org.eclipse.jdt.core.compiler.source=1.8
```
Please remove this file because it causes the default Eclipse project settings to be overridden. See also #6106.

wborn commented on bundles/ui/org.eclipse.smarthome.ui.login/.settings/org.maven.ide.eclipse.prefs:
```diff
@@ -0,0 +1,9 @@
+resolveWorkspaceProjects=true
+resourceFilterGoals=process-resources resources\:testResources
+skipCompilerPlugin=true
+version=1
```
Please remove this file because it causes the default Eclipse project settings to be overridden. See also #6106.

splatch commented:
I moved whole thing a little bit ahead - authentication part works, however AudioServlet tests are failing due to missing `HttpService` and `HttpContext` references. I need someones tip on how to solve that. PDE/Tycho test infrastructure remains unknown to me and I have no clue how to bring these two services alive for the test other than registering mocks.

splatch commented on bundles/io/org.eclipse.smarthome.io.http.auth.form/web/index.html:
```diff
@@ -0,0 +1,28 @@
+<html>
```
Form is just an example, and it doesn't work at this stage so I will remove it to not cause any confusion.

splatch commented on bundles/io/org.eclipse.smarthome.io.http.auth/src/main/java/org/eclipse/smarthome/io/http/auth/internal/AuthenticationHandler.java:
```diff
@@ -0,0 +1,160 @@
+                        return;
+                    } catch (AuthenticationException e) {
+                        failed++;
+                        logger.info("Failed to authenticate using credentials {}", credentials, e);
```
It depends on credentials type and their implementation of `toString` which needs to be made in safe way regardless of this log statement as credentials might be accidentally logged anywhere down to this path.

kaikreuzer replied:
Could you remove the `, e` here (or only log it if debug is enabled)? Currently, this fills the log with stacktraces, if somebody unsuccessfully tries to log in.

splatch commented on bundles/io/org.eclipse.smarthome.io.http/src/main/java/org/eclipse/smarthome/io/http/Handler.java:
```diff
@@ -0,0 +1,31 @@
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * A request handler.
```
It sits in http package already and handles both request and response. Some implementations might only want to work with response thus adding "request" might not be best fit. `RequestHandler` becomes then very suggestive unless you know method signatures which most people do not unless reach this type.

splatch commented on bundles/io/org.eclipse.smarthome.io.http/src/main/java/org/eclipse/smarthome/io/http/internal/DefaultHandlerContext.java:
```diff
@@ -0,0 +1,72 @@
+
+            if (!hasError && hasError()) {
+                // we didn't have an error and we have it now, meaning a current handler reported issue.
+                // so here reset cursor to 0 and restart execution forcing handleError method execution.
```
This part got changed a little bit, however there might be a case where AuthenticationHandler fails and then there is bunch of Handlers in later stage (after auth or authz) which can generate error response or challenge client with famous HTTP-Basic headers. :-) Moving error handling to Handlers makes it plugable.

kaikreuzer commented:
> however AudioServlet tests are failing 

Travis is happy, so what do you want more? :-)

kaikreuzer commented:
What is missing is your Signed-off-by footer on the commit message, though.

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/auth/SecurityException.java:
```diff
@@ -0,0 +1,51 @@
+ * @author Łukasz Dywicki - Initial contribution and API
+ *
+ */
+public class SecurityException extends RuntimeException {
```
@splatch Any comment?

splatch replied:
I am not in favor of neither checked or unchecked exceptions. In this particular case I relied on runtime exception as most of places where such exception could occur usually propagates them up to parents who dispatched requests to let them deal with it.
As I said earlier - I am not religious on any approach. I will update code to use checked exceptions which will make method signatures a bit longer but make authentication error handling explicit and not implicit.

kaikreuzer commented on bundles/io/org.eclipse.smarthome.io.http.auth/src/main/java/org/eclipse/smarthome/io/http/auth/internal/AuthenticationHandler.java:
```diff
@@ -0,0 +1,169 @@
+    }
+
+    @Reference(cardinality = ReferenceCardinality.OPTIONAL, policy = ReferencePolicy.DYNAMIC)
+    public void setAuthneticationManager(AuthenticationManager authenticationManager) {
```
typo: setAuthenticationManager

kaikreuzer commented on bundles/io/org.eclipse.smarthome.io.http.auth/src/main/java/org/eclipse/smarthome/io/http/auth/internal/AuthenticationHandler.java:
```diff
@@ -0,0 +1,169 @@
+        this.authenticationManager = authenticationManager;
+    }
+
+    public void unsetAuthneticationManager(AuthenticationManager authenticationManager) {
```
typo: unsetAuthenticationManager

kaikreuzer commented on bundles/io/org.eclipse.smarthome.io.http.auth/src/main/java/org/eclipse/smarthome/io/http/auth/internal/RedirectHandler.java:
```diff
@@ -0,0 +1,60 @@
+import org.osgi.service.component.annotations.Component;
+
+/**
+ * Handler located after authentication which redirect client to page from which he started authentication process..
```
remove duplicate  dot at the end

kaikreuzer commented on bundles/io/org.eclipse.smarthome.io.http/src/main/java/org/eclipse/smarthome/io/http/internal/HttpContextFactoryServiceImpl.java:
```diff
@@ -0,0 +1,51 @@
+    private WrappingHttpContext httpContext;
+
+    @Override
+    public HttpContext createDefaultHttpContext(Bundle bundle) {
```
Comments?

splatch replied:
As pointed in referenced comment - it is still completely legal to use yet another http context, however for consistency and unification reasons there should be one logical context. A centralized authentication/authorization handler will then take care about deciding if access to resource should be granted or not.

To counter argument openness of `AudioServlet` - callback link might contain a semi random value or token which will make sure that caller is not trying to emit inappropriate requests. Similar like you do in dropbox or google doc share links.

kaikreuzer commented on bundles/io/org.eclipse.smarthome.io.http/src/main/java/org/eclipse/smarthome/io/http/servlet/BaseSmartHomeServlet.java:
```diff
@@ -0,0 +1,68 @@
+ *
+ * @author Łukasz Dywicki - Initial contribution and API
+ */
+public abstract class BaseSmartHomeServlet extends HttpServlet {
```
Question: Instead of providing abstract base classes that need to be extended by everyone, wouldn't it be nicer to follow the whiteboard pattern and only have those implement a "SmartHomeServlet" interface, while the framework provides a service that collects those and registers them on the http service? From the past we have some bad experiences with using base classes as an API and thus would want to go for proper interfaces whereever possible. Wdyt?

splatch replied:
Difficulty with whiteboard is marrying everything together - you register things using `SmartHomeServlet` interface. In order to bind it with http service you need to make explicit cast to `HttpServlet`. It would be much better if we would have `SmartHomeServletProvider`, because then we avoid casts, however this interface would be a factory and updates of it will become harder.

We will still need to do binding of http context, the only difference is that we will do that centrally in whiteboard extender. Beyond that we will need to fetch servlet path (probably one of methods to be added in interface).
I understand your argumentation and motivation, however I wouldn't like to push that change over with this PR to do not mix things. It is definitely an improvement to above base class, however if we want to stick with this approach we need to decide what goes first - auth or whiteboard.

kaikreuzer replied:
Ok, let's not let the scope explode and leave the whiteboard refactoring for later. It just didn't look nice to have so many places in the code now extending a new abstract class although we do not want to use this pattern.

splatch commented on bundles/io/org.eclipse.smarthome.io.http.auth/src/main/java/org/eclipse/smarthome/io/http/auth/internal/AuthenticationHandler.java:
```diff
@@ -0,0 +1,160 @@
+            return false;
+        }
+
+        // TODO add decision logic so not all URIs gets secured but only these which are told to be
```
If we have 3rd party service which doesn't want to belong to "smarthome" domain context then it is free to use its own  context via `HttpService.createDefaultContext()`. If a service on other hand is part of smarthome then, I believe, it should bind with logical context provided and controlled by system.
Thanks to that unification we will have a central place to control authentication and (maybe) authorization. This method is left especially for that purpose - to leave a space for decision if authentication is required or not.

kaikreuzer commented on bundles/io/org.eclipse.smarthome.io.http/src/main/java/org/eclipse/smarthome/io/http/servlet/BaseSmartHomeServlet.java:
```diff
@@ -0,0 +1,68 @@
+ *
+ * @author Łukasz Dywicki - Initial contribution and API
+ */
+public abstract class BaseSmartHomeServlet extends HttpServlet {
```
Could you add a `serialVersionUID` to all servlets? Eclipse shows a warning that this is missing although the class is Serializable.

kaikreuzer commented on bundles/io/org.eclipse.smarthome.io.http/src/main/java/org/eclipse/smarthome/io/http/internal/SmartHomeHttpContext.java:
```diff
@@ -0,0 +1,91 @@
+ * {@link #handleSecurity(HttpServletRequest, HttpServletResponse)} method which is based on injected list of generic
+ * handlers.
+ *
+ * @author Łukasz Dywicki
```
add "- Initial contribution"

kaikreuzer commented on bundles/io/org.eclipse.smarthome.io.rest.auth/src/main/java/org/eclipse/smarthome/io/rest/auth/internal/SmartHomeRestServletConfig.java:
```diff
@@ -0,0 +1,58 @@
+        return new Hashtable<>();
+    }
+
+    @Reference(policy = ReferencePolicy.STATIC, cardinality = ReferenceCardinality.MANDATORY)
```
`@Reference` is enough, the rest is the default anyhow

kaikreuzer commented on bundles/io/org.eclipse.smarthome.io.rest.auth/src/main/java/org/eclipse/smarthome/io/rest/auth/internal/SmartHomeRestServletConfig.java:
```diff
@@ -0,0 +1,58 @@
+ *
+ * This extension doesn't do much, just forces usage of http context created somewhere else.
+ *
+ * @author Łukasz Dywicki
```
add "- Initial contribution"

kaikreuzer commented:
I've just tried to test it on the openHAB distro, but for some reason, the bundles that contain servlets are in "waiting" state as they fail to resolve a service. The service seems to be correctly provided, though:
```
openhab> bundle:diag 221
Eclipse SmartHome Basic UI (221)
--------------------------------
Status: Waiting
Declarative Services
org.eclipse.smarthome.ui.basic.internal.servlet.CmdServlet (255)
  missing references: HttpContextFactoryService
org.eclipse.smarthome.ui.basic.internal.servlet.WebAppServlet (242)
  missing references: HttpContextFactoryService

openhab> bundle:services 113

Eclipse SmartHome HTTP Interface Bundle (113) provides:
-------------------------------------------------------
[org.osgi.service.http.HttpContext, org.eclipse.smarthome.io.http.WrappingHttpContext]
[org.eclipse.smarthome.io.http.HttpContextFactoryService]
```
Any idea what might be wrong?

kaikreuzer commented:
Argh, forget my question, I found the reason:
```
221 │ Waiting  │  80 │ 0.10.0.201809271800    │ org.eclipse.smarthome.ui.basic
```
It downloaded the apps from the remote repo and didn't use my local build...

kaikreuzer commented:
I have created CQ 17756 - please refrain from touching this PR from now on!

marziman commented:
If this is merged, I want a big coffee from @kaikreuzer and @splatch for always pushing it forward 😁

kaikreuzer commented:
> I want a big coffee from @kaikreuzer and @splatch for always pushing it forward

@marziman First time you appear here on this PR and you want kudos? I think you should rather spend a round of beer for Lukasz and myself, who spent endless nights on coding, reviewing and testing!

marziman commented:
Hmm @kaikreuzer, I think you missestimate the efforts behind the scenes. That things go forward and becomes a contribution. We already have this feature running for a longer time and it would stuck without our push. So no beer this time..

marziman commented:
I dont see it as a fair argument either, cause all the meetings and discussions in the RBAC issue are blended out.But ask @splatch he knows how much dnergy went in from my side. So his kudos are sufficient for me..

splatch commented:
Gentleman, there is no need to make such arguments. All of you did outstanding job on various fronts to let this happen. Kudos belongs to all of you who supported this work.

I believe that Kai wanted to bring some fun with his comment and get a free beer as well (who wouldn't like to get one!?).

I confirm that there was an unofficial review and test phase in July. It happened two weeks before submitting this PR and was provided by @svilenvul with Mehmet push for it. Svilen reported several issues (see Code-House/smarthome#1, Code-House/smarthome#2, Code-House/smarthome#3, Code-House/smarthome#4, Code-House/smarthome#5) which I've fixed in early August.

There was no big calls with all parties, as there was very little to debate. Plumbing work had to be done anyway, and it could be verified and discussed during review as it happened. In terms of credits we also need to mention @antoniosv who worked with me earlier this year, provided some code, and tested several things related to JWT. His help and testing also moved us forward.

On bright sides, for sure there is much more work on this topic in front of us!

Cheers,
Lukasz

splatch commented:
> Could we have a short description of what feature this PR brings to the final user ?

Within few bundles/features provided within this PR you will be able to use HTTP Basic authentication to secure your installation. There is no authorization support yet (that's why this PR is called part one). In later steps we need to provide JWT handling on both - server and client side. 

> Does it introduce an authentication to connect to ESH/openHAB ? In this case, how and where do we define the authorized accounts ?
Without diving to much into details - you need to:
- property `realmName=karaf` inside `org.eclipse.smarthome.jaas` configuration to use the same user/password as you do to access openhab/esh console. 
- property `authentication.enabled=true` inside `org.eclipse.smarthome.auth` configuration
- install feature name `esh-io-transport-http-auth-basic`

First step defines how you validate user/password information.
Second step enables security.
Third one adds support to extract user/pw from incoming http request via basic auth (browser prompt).

Once authenticated you will have access to everything, otherwise you should see nothing.

Kind regards,
Lukasz

kirantpatil commented:
@splatch, do we have web interface to create, modify and delete users ?

splatch commented:
> @splatch, do we have web interface to create, modify and delete users ?

Not really, unless you write your own `AuthenticationProvider`. For openhab installations once this change get merged and configuration applied - you can use users.properties file.

kaikreuzer commented:
> I believe that Kai wanted to bring some fun with his comment and get a free beer as well

Definitely. I didn't mean to offend anyone, nor to play downgrade any contribution.

Nonetheless, I'd like to remind everyone of the rules of open source engagement: Contributions and discussions have to be public within this project, that's all that counts when it comes to attribution.

> I confirm that there was an unofficial review and test phase in July.

The much better way would have been a WIP PR at that time already - @svilenvul could have provided his feedback as a review on that PR.

> Svilen reported several issues which I've fixed in early August.

Sorry, but that's a good example how NOT to do open source: Those issues have no comments, they do not directly reference any code/commit, they are still open and their status is completely unknown. Again, please do WIP PRs in future instead, that would make much more sense for everyone.

> Could we have a short description of what feature this PR brings to the final user ?

@lolodomo In short, it does not bring anything to the final user as it is only a framework feature that can potentially be used and integrated into ESH solutions. So once this is merged, a next step for me would be the integration in openHAB - I briefly tested this already, but still had issues, so it can still take a while before anything becomes available to openHAB users. But for this, let us open an issue at openhab-distro instead for further discussion.

marziman commented:
@kaikreuzer,

I think the people know well how OSS works. The things we implement is not intended to be opensource atm. So went the right way as we made issues in the github of @splatch, where the current state was (visible to the public also).

So I think it is not about anyone getting kudos. It is about accepting the fact that it is more than just submitting a PR. 
Again, we pushed this stuff and that fact shouldnt be made smaller as it was.

I think instead of trying OSS process dogma, lets stay at the core of that discussion: Respecting all involved peoples work.

BR Mehmet

kaikreuzer commented:
> The things we implement is not intended to be opensource atm.

Is this PR a part of the work that "we" (i.e. you, whoever you exactly refer to here) are working on?

Just to be clear: If there is anyone but @splatch that had a significant share of the contribution of this PR, I will have to withdraw the CQ immediately as this isn't correctly stated.

splatch commented:
> Just to be clear: If there is anyone but @splatch that had a significant share of the contribution of this PR, I will have to withdraw the CQ immediately as this isn't correctly stated.

Entire code contained within this PR was written by me and no one else. In terms of code there are no contributions from other people.

kaikreuzer commented:
Good news, everyone, the CQ has already been approved for check-in!
I'll do a new ESH stable soon, so we can then see how to get this working in openHAB for a start.
At the same time, we should start working on the "part 2", i.e. OAuth2 & JWT support as discussed.

kaikreuzer commented:
> so we can then see how to get this working in openHAB for a start.

I have created https://github.com/openhab/openhab-core/pull/412 for this.

kirantpatil commented:
May I know how to access basic auth using API ?

splatch commented:
> May I know how to access basic auth using API ?

@kirantpatil You need to install appropriate feature and then turn on authentication (`org.eclipse.smarthome.auth.cfg` with `authentication.enabled=true`). Second step is enabling of `AuthenticationProvider` who will verify username/password combination - this is yet another step which is specific to product you use.
If you are interested in this topic in context of openhab, please start forum thread at http://community.openhab.org/.

Cheers,
Łukasz

lblabr commented:
is there already a thread in community, i can't find one related to this PR, could somebody give me a hint?

thanks in advance
